### DISCUSSION TITLE
Code rollback

### DISCUSSION THREAD

**Discussion by vosen** [vosen/ZLUDA#261] [General]
I understand that the topic might generate strong opinions, but please refrain from:
* using profanities
* posting the links to forks or other means for getting the code

**Comment by ethanc8** [comment]
Are you able to share the email that AMD originally sent you giving you permission to release ZLUDA 3?

**Reply by vosen** [reply]
I'll check with the sender

**Reply by vosen** [reply]
Sorry, but the sender is not allowing me to share the email publicly

**Reply by szmarczak** [reply]
@vosen What is the position of the person who gave you the permisison? And what is what is the position of the person who reverted the decision? Also what did the permission allow you to do, in exact words?

**Reply by vosen** [reply]
Sorry, but I have to stop that line of inquiry. The issue of sender and email content is ultimately meaningless here: AMD legal acknowledges that it wa sent, but they believe no matter its content, it is not legally binding

**Reply by szmarczak** [reply]
> The issue of sender

No. It depends on their legal power. If the CEO said that, then most likely it would be binding (unless in an unofficial setting). However, if a support guy said that someone else said it's ok, then most likely it wouldn't be binding.

> no matter its content, it is not legally binding

That's definitely wrong. Imagine you signed a contract and then the other party says "no matter its content it is not legally binding". That's not how it works. If a person who has the right to give permissions says you have one, then you have one.

On a side note, covering up tracks by deleting links to MIT-licensed software is kind of a dick move. However you're free to do so since it's your repository.

**Reply by ethanc8** [reply]
Earlier you said that your employment contract gave you permission to release the code if AMD is no longer interested in the code:
> One of the terms of my contract with AMD was that if AMD did not find it fit for further development, I could release it.
(from the README.md of Version 3)

Is AMD now interested in the code, or was that not really in the contract and rather was some kind of verbal agreement?

**Reply by EliasTheGrandMasterOfMistakes** [reply]
> Sorry, but I have to stop that line of inquiry. The issue of sender and email content is ultimately meaningless here: AMD legal acknowledges that it wa sent, but they believe no matter its content, it is not legally binding

lawers being lawers

**Reply by oleid** [reply]
>  And what is what is the position of the person who reverted the decision? Also what did the permission allow you to do, in exact words?

This whole situation only makes sense to me if AMD's lawyers suddenly have a strong incentive to "take care" of this repository, i.e. if they got a not-so-nice letter from Nvidia's lawyers. Of course this is purely speculation.

**Reply by erkinalp** [reply]
Are you considering to seek for a formal reinstatement of the licence for the AMD code?

**Reply by wiki-me** [reply]
> Earlier you said that your employment contract gave you permission to release the code if AMD is no longer interested in the code:

Yeah if that is true then retracting the code sounds illegal and unethical , maybe the [Software freedom law center](https://en.wikipedia.org/wiki/Software_Freedom_Law_Center) could be consulted, the free software foundation and the GNOME/KDE foundation might also help.

**Reply by erkinalp** [reply]
@vosen If you manage to formally restore your permission to release this code, is it just a rollback away, or do you have further changes you privately made?

**Reply by sempervictus** [reply]
> Sorry, but the sender is not allowing me to share the email publicly

Could you please share the spam filter logs or MTA logs from the third parties which do have copies since those are not the original correspondence but intermediate provider data about the emails sent and received between the TO/FROM fields in the headers? When i dealt with the prior incarnation of the "graphics legal team" nearly 30y ago, they had not figured out email encryption to avoid that loss of confidentiality and something tells me that they still have not given how they're handling this.

Any party with access to your mailbox is also not beholden to the terms of the work-for-hire contract under which the compensation for this was presumably structured - quite likely that one or two tech-pub journalists can be found who can be trusted not to abuse access to the mailbox(es) containing this thread and incidentally may find a very wide and animated audience for the contents of that specific thread. Same goes for backup providers, IT people, etc - unless the contract stipulated TS-grade security measures for work product, communications, and systems; the other party cannot have any expectation of confidentiality given their own well-publicized abysmal security practices (and products) despite the immense budget and talent pool available at their level.

**Reply by RealVarHar** [reply]
The worst thing is, nvidia will not legally shut this down, because of a risk of a really mean antitrust legislation. 
But since its amd shooting its own foot this project will likely not survive this setback.

**Reply by DJHanceNL** [reply]
you guys dont get it. No AMD is not interested in zluda all of a sudden.  Zluda makes us use cuda. even if @vosen made a perfect piece of software. it's using nvidia trash. They want their own cuda. They thought zluda was okish, but not ready for real world use for guys all around the world making nsfw ai content. So it was harmless.  

Now they see it works pretty great, with some tweaks. and even me as a non programmer can use it for stable diffusion. So they want it gone. They want to release their own alternative. And for you and me, to early adopt their thing. So they could start competing with nvidia and their golden boy CUDA. 

So zluda release was ok if it was shit, now it's pretty decent they dont want us to use it.

**Reply by TinfoilyTim** [reply]
Why wouldn't amd want the code to be out tho? I know they are working on their own thing but this closes the gap just a little bit hardware wise. Nvidia have already been in trouble because of Cuda so they most likely won't go after amd, so I don't see a reason they need to take this down.

**Comment by powerpuffboysz** [comment]
Is gamework support completely shot or will it just take a long time to rebuild?

**Reply by vosen** [reply]
Well, if I could do this, then someone sufficiently motivated could do this too. The problem is that any org that is interested in funding ZLUDA is not interested in funding it for GameWorks

**Reply by Koth-87** [reply]
> Well, if I could do this, then someone sufficiently motivated could do this too. The problem is that any org that is interested in funding ZLUDA is not interested in funding it for GameWorks

Is it possible to crowdfund you to rebuild it? I know that might be asking a lot...

**Reply by vosen** [reply]
It's not the matter of money, but rather time. I need to prioritise time spent on things for which there's a business case. Doesn't mean I can't take a quick detour to enable an easy workload, but GameWorks is neither quick nor easy

**Comment by qwertychouskie** [comment]
Would be nice to see CUDA implemented on top of Vulkan Compute so it's truly cross-platform and cross-vendor, though I'm not sure what the logistics are there.

**Reply by vosen** [reply]
Vulkan Compute is missing features. More detailed answer here: https://github.com/vosen/ZLUDA/discussions/115#discussioncomment-8596605

**Reply by mirh** [reply]
SYCL is already aspirationally meant to be "truly cross-platform and cross-vendor".

Besides, if you mentioned Vulkan because your purpose was gaming-related (physx) that would still be pretty hard considering everything is win32 based. But that's no biggie, really, considering wine-staging can pass through just about any call that is available to the native system (ironically including CUDA itself btw). So there's no problem here.

**Reply by vosen** [reply]
I wrote tens of thousand lines of SYCL code and my opinion is that it's not going to challenge CUDA at all.
Wine can pass calls, but for gaming specificially it might be more complex (depending if it's original PhysX, GameWorks or DLSS). In the case of gaming, most of the time you rely on CUDA <-> DX interop. To implement this on AMD HW you need to have HIP <-> DX interop too. Interop boils down to HIP having a knowledge of internal layout of various DX data structures and understanding its internal operations to inject its own data/operations.
HIP <-> DX works (just barely) on Windows, but on Linux HIP <-> VK is not going to be very useful, because AFAIK everyone uses Mesa and not AMDVLK. And the fact that you are going throught adds another layer (because you don't want to interop with DX object but underlying Vulkan one).

**Comment by Koth-87** [comment]
I am absolutely heartbroken :(

I know this probably trivial compared to what other people were gonna use it for, but Arkham Knight is my favorite game and I was dying to try this out. I even commented on it a few weeks ago: https://github.com/vosen/ZLUDA/discussions/31#discussioncomment-10056834

Truly a sad day :(

**Comment by niansa** [comment]
How can we donate to show our support as simple users?

**Reply by teamolhuang** [reply]
Really needs a mean of donation for us to support ZLUDA. I will not choose AMD for my future devices/replacements again since this is such an unacceptable move, pulling the rug under the creator worked for them.
But before my current card dies, I still need to generate my anime wife images, and your work has always helped me a lot. Please let us support your magnificent work.

**Reply by vosen** [reply]
Thanks, I've been in contact with other, longer-running open source projects of similar size and my conclusion is that the best way forward is funding by an organization. Which I'm in the process of finalizing, so you can buy yourself a daki or a fig

**Comment by BaalNetbek** [comment]
That was awesome run. I hope this hit doesn't set you down for long.

**Comment by sempervictus** [comment]
If anyone from AMD is reading this thread, i am `RageLtMan` of the ancient Rage3D ATI driver project.
I told ATI's people at the turn of the century that if they did not implement unified drivers (thankfully that went though but not for the codebase we were manhandling) and a language abstraction for GPGPU good enough to beat the competition, then Nvidia would eat ATI's lunch. They did, then AMD ate the starved to death carcass.
**Please stop making ATIs mistakes again from beyond its grave** and _embrace the risks of battling for market share_ especially when you wage that battle by FOSS proxy; because your hardware quality has no bearing on adoption if CSPs and AI/HPC scalers won't buy it (as their users expect CUDA, various NV interfaces, etc). Its more cost effective to play the legal game if it comes to that and spend 0.1% of the profit to be gained in the HPC/AI market dragging out the process while gaining market share than to keep "playing it safe" (from a 2nd place position so far behind 1st that its barely relevant). The legal department is rarely the one to effect market share acquisition... engineers who take _calculated_ risks are better decision makers in a dynamic market than people prescriptively fearful of the consequences whether they fail or succeed. Source: these days, run a consultancy building/securing/orchestrating these fabrics within the scalers and their HPC peers where AMD kit is almost nowhere to be seen. This market needs legitimate competition to ensure ongoing innovation of capability vs mere increases in complexity and throughput of packaging.

**Reply by erkinalp** [reply]
@saadrahim

**Reply by sempervictus** [reply]
> @saadrahim

I'm happy to take my case all the way to Dr Su if she can find the time and bandwidth to speak.

**Comment by roastedpeacock** [comment]
Thank you for developing ZLUDA!

I think it is unfortunate this project had to be rolled back. Understood if you cannot comment but I would push back harder on censorship of this project. I hope a correct decision will be made eventually. 

In any case continuing research into vendor-agnostic GPU computation is great. However I fear any further effort on your part would be on an unequal footing considering the existing code is already out there and a potential fork by third-parties is theoretically possible given enough skill and time. Not going to go into specifics however previous open-source projects that have ran into questionable legal territory have been forked and continued by unaffiliated developers.

**Reply by vosen** [reply]
I'm not worried about forks. It's not a competition

**Reply by erkinalp** [reply]
do you plan to pursue legal avenues to restore the AMD version (v3), or would it just be too much effort for too little gain?

**Reply by vosen** [reply]
I won't. It's too expensive (think hundreds of thousands of dollars) with no guarantee of success

**Reply by wiki-me** [reply]
Did you talk to the organisation that wants to sponsor the project? writing in a contract you can open the code and saying its not legally binding sounds like a slam dunk case to me. They might have a lawyer on retainer so the cost to them is basically zero and they might be able to do a better risk benefit analysis because they are lawyers.

**Reply by vosen** [reply]
Yes, they are aware of all the details. I talked to a lawyer. It's not a mega-corporation, the costs would really hurt

**Reply by vosen** [reply]
And one more thing: it was not written explictly into the contract, but as a separate email. I know that the previous README mentioned it being part of the contract, but I was trying to write a good explanation without being pedantic

**Reply by wiki-me** [reply]
oh well, at least we learnt a lesson i guess, one of the [conditions](https://neathousepartners.com/blog/is-an-email-a-legal-document) for having an email become legally binding is apparently "A clear intention to be legally bound by the contract". I just hope with the new organisation the agreement to open source is in the contract.

**Reply by erkinalp** [reply]
next time have your employment contract notarised

**Comment by dbermond** [comment]
The latest version is still  being shown as `v3`, which contains the AMD code published some months ago.

Shouldn't the latest version also be reverted back to `v2`? Will this AMD code, which is still present in the online repository history, be somehow rebased/removed?

**Reply by roastedpeacock** [reply]
The newest commit in my checkout is below

`9e56862 (Update README with CGBN information (#237), 2024-05-23)`

**Reply by vosen** [reply]
Hmmm, I'll leave the binary for now. I've got the request to remove the source code. Also please don't "post the links to forks or other means for getting the code"

**Reply by erkinalp** [reply]
in github's perspective, forks are branches of the same repository, not separate remotes like how forks work on other SCMs
so you should request a detachment or a takedown of other forks to truly break the connection

**Reply by vosen** [reply]
Yes, but I've been asked to rollback the code in the repo, not take down every trace of the project and its code from the internet. Let's leave it be

**Reply by erkinalp** [reply]
then, a detachment (email github support to request all forks of ZLUDA to be detached from you), so you won't be seen as the perpertrator in github's eyes

**Comment by phly95** [comment]
I guess I won't be buying AMD for my next PC. To not be able to games from 2008 (eg. Mirror's Edge) at max settings on a 2024 card is getting ridiculous. People do play old games, and will still choose Nvidia if it means they can play those old games at max settings.

Congrats AMD, you just cost yourselves billions of dollars in lost sales. (BTW, I have never bought AMD because of that slowdown in Mirror's Edge that happened when I was a kid and will not until it is fixed)

In fact, this is probably what almost drove AMD bankrupt until they were saved by the game consoles and then Ryzen. If AMD supported PhysX and Gameworks features, people wouldn't think of AMD as the scraps you buy because Nvidia is too expensive.

**Reply by oleid** [reply]
> To not be able to games from 2008 (eg. Mirror's Edge) at max settings on a 2024 card is getting ridiculous

You mean because of PhysX? Other than that, I don't see why the game shouldn't run fast on today's GPUs.
It worked on mobile GPU 11 years ago: https://www.youtube.com/watch?v=scg9HQHwgpI

**Reply by phly95** [reply]
Yes, because of PhysX. It runs fine at first, but if you have AMD, once the glass is smashed, your frame rate is tanking. And ofc, you aren't running at max settings unless you've enabled PhysX. I have no idea if any other games have this kind of issue. Yes, Nvidia is to blame for the issue (limiting non-GPU PhysX to a single CPU core is clearly anti-competitive, given a modern CPU can easily run PhysX if that limitation did not exist), but AMD is actively preventing a fix now. That game is certainly not getting a PhysX update now unless there's a remaster.

**Reply by vosen** [reply]
I'm sad about GameWorks being gone too.
I can't comment on the Mirror's Edge CPU GameWorks, but my experience from enabling CUDA GameWorks is that it does some nonsensical things that are buggy, but just happen to work

**Comment by marcussacana** [comment]
So the project keep alive, you will just rewrite it...?
how much time you think will be needed to be at the same stability level of the current legacy version?

**Reply by AssassinK786** [reply]
This is Better option for everyone
Just ask for some extension **@vosen !**

**Reply by vosen** [reply]
Project is alive. you can see progress of the first big PR here: #267
I'm aiming for one year for the same approximate level of functionality. Meaning applications that work right now will work in as well (as long as thay are in the scope)

**Comment by TinfoilyTim** [comment]
Will we always be able to download the pre-takedown release?

**Reply by erkinalp** [reply]
there are forks we're not supposed to name here, feel free to look at them

**Comment by DJHanceNL** [comment]
This is pretty standard practice, right? Even if it works great, CUDA is NVIDIA's golden ticket, and they don’t want that to change. The fact that you and I can make Stable Diffusion and some other AI programs work almost as well on AMD GPUs, which cost half as much as NVIDIA's, is great for us, but not for AMD. We are still using something that hooks into the competitor's golden ticket. Eventually, AMD will want to have their own CUDA equivalent. If they have something like that in the works and it's almost ready, they need to steer us away from ZLUDA and get early adopters to jump on their new solution for it to take off. They might try to compete with CUDA and NVIDIA.

So, even though @vosen did an amazing job with ZLUDA, I think it was always **Plan B** to have a way of running CUDA on AMD. Even if @vosen made the impossible work and ZLUDA would be complete and run perfectly, AMD would stil choose **Plan A over Plan B.** Plan A involves having their own CUDA alternative that they can push, own, and require their hardware for, which means more money. Companies always choose Plan A, even if it’s worse, flawed, or completely stupid.

**Comment by phly95** [comment]
Would it be possible to open source the Gameworks support code, even if it doesn't do anything useful or is completely broken? It might be useful for other developers interested in studying how Gameworks works, since it doesn't seem working Gameworks support will be a priority for a long time coming.
It shouldn't be a problem if it doesn't include the AMD code needed for it to work, given the Gameworks code was made after leaving AMD. Perhaps it could be used to make a non-CUDA Gameworks framework in Vulkan or something.

**Comment by dobi70** [comment]
Hi. I might have a stupid question but maybe vosen or someone else can help/answer it.
First things first. I do not game at all. I use my PC for video editing, maybe colorizations with pixbim in the future....
I just returned a Rx 7600 xt after finding out that zluda was removed from here. I am an Intel CPU guy but always an ATI/AMD GPU person.
Now here is my question. I don't have a golden goose. 
Should I get a Rx 7700 xt or a Rtx 4600 ti 16 gb. 
I don't know when zluda or anything like it will come back that can run cuda on AMD.
Is there something else that can efficiently run cuda on AMD? When will zluda return? 
Days? Months? Years? Or never?
What is AMD up to, to request zluda be taken down? I hate to blow money out on something I'll regret later, but I also don't want to support Nvidia's monopoly game.
Any input would be appreciated

**Reply by marcussacana** [reply]
Him said that his objective is to get zluda in the same level of the current version with one year.
If you have money to spent should just get the overpriced RTXs, otherwise AMD.

**Comment by nitefood** [comment]
I may definitely be way off in my understanding of what's at stake here, and in that case, I apologize in advance. I am not trying to downplay the importance of what's going on here.

But have you considered taking the repo private and switching to a binary only release model? I know this means effectively turning to the dark side of closed source, but then again perhaps there could be a solution to grant access to the repository to contributors that are willing to sign an NDA, or something?

I mean if what they asked of you was to _make the source code repository **(and nothing else)** not publicly available anymore_, one may speculate that once you make it private, and cherry pick who is allowed access to it, all the while keeping the binaries up, it could be enough to make them happy...

**Reply by mirh** [reply]
Have you considered checking the sources and noting that 15K lines of code were merged just last week?
No looking back now.

**Reply by vosen** [reply]
Keeping the source closed would not help. AMD position is that it's their intellectual property, taking it closed solurce would not resolve that claim. It's better to roll back than to fight it

**Reply by nitefood** [reply]
> Keeping the source closed would not help. AMD position is that it's their intellectual property, taking it closed solurce would not resolve that claim. It's better to roll back than to fight it

I see your point. It's always sad when open and brilliant software has to get censored because a mega corporation can't take the right stance and support the community, let alone listen to it. Best of luck for your future work.

**Reply by erkinalp** [reply]
*meanwhile the largest fork quietly adding ROCm 6 support* (not going to link to them, but you can easily find them using github search)

**Comment by allzhukov** [comment]
How to run zluda on intel?

**Comment by Salleyi75** [comment]
Hello, this is not related to the "code rollback" topic, i just wanted to ask this question since you are an expert and happened to create Zluda. For the past 2 months i have been wondering what exactly is making people say Nvidia is better than Amd for ai (i am contemplating on buying some hardware to run local ai, plus am a programmer so i don't mind error messages here and there), my initial assumption was that Nvidia has some special hardware that works for ai that amd lacks, so i delved a bit into cuda programming and hip programming, i mean they seem to achieve the same goals in different ways, but they seem equal in my eyes, what exactly is AMD missing, looking at the fact that Zluda exists, tells me that amd hardware is good enough because if the hardware was not good enough i dont think there will be a point in creating zluda, so my final conclusion is, people are used to programming cuda than hip/rocm. Kind of link the ps3 cell processor situation repeating itself where people were not familiar with the style of programming on the cell processor hence they did not really want to make games for the ps3 and if they did, it was badly optimized, is that what is happening to amd ?

**Reply by vosen** [reply]
Consider this: all major desktop GPU vendors share lineage: sure they are used for compute now, but they evolved to serve gaming APIs. As such, most of their capabilities overlap. We see more and more compute-only hardware features and removal of certain graphics HW from compute cards, but the baseline is very very similar. They program similarly. Even NV and AMD ISAs are similar (Intel is a bit of an outlier).
I think the NVIDIA is the first to add new HW features. They range from genuinly useful (various ML low precision instruction) to harmful (independent thread scheduling). New HW features sometimes appear in Intel/AMD GPUs bit later, sometimes not at all, sometimes earlier, but if you are just starting GPU programming they will be quite esoteric.

The big divide is software. For NV the whole stack (compiler, runtime, profiler, libraries, debugger, etc). mostly works and works as you'd expect it to. You've asked specifically about AMD, so I'll talk specifically about AMD. They had the right idea: provide a GPU programming language and libraries that are sufficiently similar to CUDA that you can just rename the functions and have yur program "just work". The problem is the execution of that idea. The list of issues with their SW is very long. Let's talk just creme de la creme. When you are compiling a HIP program for an AMD GPU you have no forward-compatiblity. Meaning, if you compile a HIP program today, it will not work with the next generation of GPUs. HIP program written today will not work on AMD Radeon RX 8XXX. This is the sort of problem that does not happen anywhere else. Not on NV GPUs (compiles to PTX, which is a forward-compatible IR), not on Intel GPUs (compiles to SPIR-V, which is a forward-compatible IR), not on CPUs (new CPU generations are backward-compatible on the ISA level).

That's not all. When you compile for AMD GPU, you have to compile for each _card model_ separately. For example RDNA 2 consists of gfx1030, gfx1031, gfx1032, gfx1033, gfx1034, gfx1035, gfx1036 (see here: https://llvm.org/docs/AMDGPUUsage.html). If you compiled just after RX 6900 XT (gfx1030) released, remember to recompile when RX 6700 XT (gfx1031) releases! And again when all those other gfx103* drop!

There is some progress with that particular issue (https://llvm.org/docs/AMDGPUUsage.html see "Generic Processors", https://www.phoronix.com/news/LLVM-AMDGCN-Flavored-SPIR-V), but I don't think it filtered already to the release stack. And it's just one of many, many issues.

**Reply by Salleyi75** [reply]
Oh ok, i see, makes sense, so i guess one of the issues is interoperability with their own hardware. Ok, thanks for the explanation, i wanted to decide between the two vendors before investing anytime. As a programmer I don't mind dealing with issues on the C and C++ level, like pointers, memory leaks, and allocation issues, as much as I am brave enough to tackle those issues, I am currently not brave enough to deal with LLVM, tried learning it in other to help with the open source zig programming language, but I gave up, maybe next year. Thanks for the insight at least I have learnt about the hardware enough to understand why ISA changes can be an issue. I think i will be able to further delve into this when I understand the LLVM IR.

**Comment by Quackdoc** [comment]
Is there a "current state" of the project? I am a bit confused on where exactly the project is compared to the past.

**Reply by marcussacana** [reply]
Him said it may take at least one year to the project be at the same level of the legacy version, it was just just 2 months since then, you will need to be patient

**Reply by vosen** [reply]
I plan to have an announcement about current state and project goals next week. Stay tuned.

**Reply by vosen** [reply]
Announcement here: https://vosen.github.io/ZLUDA/blog/zludas-third-life/
